hazelcast/src/main/java/com/hazelcast/osgi/impl/Activator.java

```diff
@@ -45,17 +45,28 @@ public void start(BundleContext context)
         // Try to start javax.scripting - JSR 223
         activateJavaxScripting(context);
 
-        if (Boolean.getBoolean(HAZELCAST_OSGI_START)) {
-            hazelcastInstance = Hazelcast.newHazelcastInstance();
+        // TODO Is it possible that OSGI service is not null (deactivated) on start?
+        if (hazelcastOSGiService != null) {
+            // If there is an active Hazelcast OSGI service leftover, deactivate it.
+            hazelcastOSGiService.deactivate();
+            hazelcastOSGiService = null;
         }
+
+        // Create a new Hazelcast OSGI service with given bundle and its context
+        hazelcastOSGiService = new HazelcastOSGiServiceImpl(context.getBundle());
+
+        // Activate new created Hazelcast OSGI service
+        hazelcastOSGiService.activate();
     }
 
     @Override
     public void stop(BundleContext context)
             throws Exception {
-        if (hazelcastInstance != null) {
-            hazelcastInstance.shutdown();
+        if (hazelcastOSGiService != null) {
+            hazelcastOSGiService.deactivate();
+            hazelcastOSGiService = null;
         }
+        // TODO Is it possible that OSGI service is null (not activated) on stop?
```

Comment:
Even though this case seems not possible by OSGI container, instead of TODO, I will throw `IllegalStateException`